### PR TITLE
Introduce a FloorPlanner trait and integrate it into the Circuit trait

### PR DESCRIPTION
**Pull request by str4d**
This is the beginning of the process to enable full floor planning capabilities in `halo2`. For now, all that a floor planner can do is synthesize a circuit, which makes it no more powerful than a layouter, but easier to use (as moving to a multi-pass layouter no longer requires changes to `Circuit::synthesize`).

### CODE REVIEW

**Review by ebfull — APPROVED** [review]

**Comment by str4d** [comment]
[Rebased to fix a doc test](https://github.com/zcash/halo2/compare/7383a8de9c068fd4f7792a611837ac2782d90f2e..6d0017f47c0119e0b6d6d671a38d5398a427b8ae).

**Comment by ebfull** [comment]
re-ACK

File: src/plonk/circuit.rs [hunk 1]
```diff
@@ -509,12 +510,38 @@ pub trait Assignment<F: Field> {
...
+///
+/// The floor planner is chip-agnostic and applies its strategy to the circuit it is used
+/// within.
+pub trait FloorPlanner {
+    /// Given the provided `cs`, synthesize the given circuit.
+    ///
+    /// Internally, a floor planner will perform the following operations:
+    /// - Instantiate a [`Layouter`] for this floor planner.
+    /// - Perform any necessary setup or measurement tasks, which may involve one or more
+    ///   calls to `Circuit::default().synthesize(config, &mut layouter)`.
```
Comment (daira) [inline src/plonk/circuit.rs:523]: This should have been `Circuit::without_witnesses().synthesize(...)`.

**Review by daira — COMMENTED** [review]
Post-hoc ACK modulo [comment](https://github.com/zcash/halo2/pull/305/files#r656638620).

File: src/plonk/circuit.rs (hunk 1 above)
Comment (str4d) [reply src/plonk/circuit.rs:523]: Ah yes, I forgot to update this comment. I'll open a separate PR.

**Pull request merged**